### ISSUE TITLE
[PP] PP report loss=-1 when LOG_RANK=gpu_nums -1

### ISSUE THREAD

**Issue by aahehehe** [pytorch/torchtitan#1232]
I training with  pipeline_parallel_degree = 16 in 2 nodes 16gpus.I had set LOG_RANK=0,15,but it still report loss=-1.
@kashif @tianyu-l @lessw2020 Could you cc?

```
+ NGPU=8
+ export LOG_RANK=0,15
+ LOG_RANK=0,15
+ CONFIG_FILE=./torchtitan/models/llama3/train_configs/llama3_70b.toml
+ overrides=
+ '[' 0 -ne 0 ']'


[rank0]:[titan] 2025-05-28 21:11:37,626 - root - INFO - Trainer is initialized with local batch size 2, global batch size 2, sequence length 4096, total steps 10 (warmup 2).
[rank0]:[titan] 2025-05-28 21:11:37,626 - root - INFO - Training starts at step 1.
[rank0]:[titan] 2025-05-28 21:13:37,816 - root - INFO - step:  1  loss: -1.0000  memory: 36.90GiB(46.63%)  tps: 4  tflops: 0.90  mfu: 0.29%
[rank0]:[titan] 2025-05-28 21:13:37,816 - root - INFO - Synchronizing and adjusting timeout for all ProcessGroups to 0:01:40
[rank0]:[titan] 2025-05-28 21:14:39,054 - root - INFO - step:  2  loss: -1.0000  memory: 42.78GiB(54.06%)  tps: 8  tflops: 1.81  mfu: 0.58%
[rank0]:[titan] 2025-05-28 21:15:40,293 - root - INFO - step:  3  loss: -1.0000  memory: 42.78GiB(54.06%)  tps: 8  tflops: 1.81  mfu: 0.58%
[rank0]:[titan] 2025-05-28 21:16:41,533 - root - INFO - step:  4  loss: -1.0000  memory: 42.78GiB(54.06%)  tps: 8  tflops: 1.81  mfu: 0.58%
[rank0]:[titan] 2025-05-28 21:17:42,765 - root - INFO - step:  5  loss: -1.0000  memory: 42.78GiB(54.06%)  tps: 8  tflops: 1.81  mfu: 0.58%
[rank0]:[titan] 2025-05-28 21:18:43,999 - root - INFO - step:  6  loss: -1.0000  memory: 42.78GiB(54.06%)  tps: 8  tflops: 1.81  mfu: 0.58%
[rank0]:[titan] 2025-05-28 21:19:45,234 - root - INFO - step:  7  loss: -1.0000  memory: 42.78GiB(54.06%)  tps: 8  tflops: 1.81  mfu: 0.58%
[rank0]:[titan] 2025-05-28 21:20:46,459 - root - INFO - step:  8  loss: -1.0000  memory: 42.78GiB(54.06%)  tps: 8  tflops: 1.81  mfu: 0.58%
[rank0]:[titan] 2025-05-28 21:21:47,679 - root - INFO - step:  9  loss: -1.0000  memory: 42.78GiB(54.06%)  tps: 8  tflops: 1.81  mfu: 0.58%
[rank0]:[titan] 2025-05-28 21:22:48,907 - root - INFO - step: 10  loss: -1.0000  memory: 42.78GiB(54.06%)  tps: 8  tflops: 1.81  mfu: 0.58%
[rank0]:[titan] 2025-05-28 21:22:48,922 - root - INFO - Sleeping 2 seconds for other ranks to complete
[rank0]:[titan] 2025-05-28 21:22:50,924 - root - INFO - Training completed
[rank0]:[titan] 2025-05-28 21:22:53,972 - root - INFO - Process group destroyed.
```

**Comment by aahehehe** [comment]
I'm running GPipe, and how can i log rank15?Could you cc ?@lessw2020 Thx!

This is config
```
[job]
dump_folder = "./outputs"
description = "Llama 3 70B training"

[profiling]
enable_profiling = false
save_traces_folder = "profile_trace"
profile_freq = 100

[metrics]
log_freq = 1
enable_tensorboard = false
save_tb_folder = "tb"

[model]
name = "llama3"
flavor = "70B"
tokenizer_path = "./assets/tokenizer/original/tokenizer.model"
# converters = ["float8"]

[optimizer]
name = "AdamW"
lr = 1.5e-4
eps = 1e-8

[lr_scheduler]
warmup_steps = 2  # lr scheduler warm up, normally 20% of the train steps

[training]
batch_size = 2
seq_len = 4096
max_norm = 1.0  # grad norm clipping
steps = 10
compile = false
dataset = "c4"

[parallelism]
data_parallel_replicate_degree = 1
data_parallel_shard_degree = 1
tensor_parallel_degree = 1  # 8-way TP
pipeline_parallel_degree = 16
context_parallel_degree = 1

[checkpoint]
enable_checkpoint = false
folder = "checkpoint"
interval = 500
model_weights_only = false
export_dtype = "float32"
async_mode = "disabled" # ["disabled", "async", "async_with_pinned_mem"]

[activation_checkpoint]
mode = "full"

[float8]
enable_fsdp_float8_all_gather = false
precompute_float8_dynamic_scale_for_fsdp = false
filter_fqns = ["output"]
```

This is all log
```
+ NGPU=8
+ export LOG_RANK=0,15
+ LOG_RANK=0,15
+ CONFIG_FILE=./torchtitan/models/llama3/train_configs/llama3_70b.toml
+ overrides=
+ '[' 0 -ne 0 ']'
+ TORCHFT_LIGHTHOUSE=http://localhost:29510
+ PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True
+ TORCHFT_LIGHTHOUSE=http://localhost:29510
+ torchrun --nnodes 2 --nproc_per_node=8 --node_rank 0 --master_addr 51.62.15.15 --master_port 29500 --local-ranks-filter 0,15 --role rank --tee 3 -m torchtitan.train --job.config_file ./torchtitan/models/llama3/train_configs/llama3_70b.toml
W0528 21:11:02.826000 4070231 site-packages/torch/distributed/run.py:792] 
W0528 21:11:02.826000 4070231 site-packages/torch/distributed/run.py:792] *****************************************
W0528 21:11:02.826000 4070231 site-packages/torch/distributed/run.py:792] Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed. 
W0528 21:11:02.826000 4070231 site-packages/torch/distributed/run.py:792] *****************************************
[W528 21:11:09.949464584 socket.cpp:204] [c10d] The hostname of the client socket cannot be retrieved. err=-3
[rank0]:[titan] 2025-05-28 21:11:19,042 - root - INFO - Starting job: Llama 3 70B training
[rank0]:[titan] 2025-05-28 21:11:19,042 - root - INFO - [GC] Initial GC collection. 0.00 seconds.
[rank0]:[titan] 2025-05-28 21:11:22,087 - root - WARNING - ENV[TORCH_NCCL_ASYNC_ERROR_HANDLING] = 1 will be overridden to 3 based on job config
[rank0]:[titan] 2025-05-28 21:11:22,091 - root - INFO - Building 1-D device mesh with ['pp'], [16]
[W528 21:11:22.859249799 socket.cpp:204] [c10d] The hostname of the client socket cannot be retrieved. err=-3
[W528 21:11:24.574823994 socket.cpp:204] [c10d] The hostname of the client socket cannot be retrieved. err=-3
[W528 21:11:24.626346075 socket.cpp:204] [c10d] The hostname of the client socket cannot be retrieved. err=-3
[W528 21:11:24.630070446 socket.cpp:204] [c10d] The hostname of the client socket cannot be retrieved. err=-3
[W528 21:11:24.633556078 socket.cpp:204] [c10d] The hostname of the client socket cannot be retrieved. err=-3
[W528 21:11:24.634119348 socket.cpp:204] [c10d] The hostname of the client socket cannot be retrieved. err=-3
[W528 21:11:24.648794344 socket.cpp:204] [c10d] The hostname of the client socket cannot be retrieved. err=-3
[W528 21:11:24.660682779 socket.cpp:204] [c10d] The hostname of the client socket cannot be retrieved. err=-3
[rank0]:[titan] 2025-05-28 21:11:33,768 - root - INFO - TikTokenizer built: #words 128256, BOS ID 128000, EOS ID 128001
[rank0]:[titan] 2025-05-28 21:11:34,111 - root - INFO - Building llama3 70B with TransformerModelArgs(_enforced='This field is used to enforce all fields have defaults.', dim=6144, n_layers=60, n_heads=64, n_kv_heads=8, vocab_size=128256, multiple_of=4096, ffn_dim_multiplier=1.3, norm_eps=1e-05, rope_theta=500000, max_seq_len=4096, depth_init=True, use_flex_attn=False, attn_mask_type='causal', eos_id=128001)
[rank0]:[titan] 2025-05-28 21:11:34,552 - root - INFO - CUDA capacity: NVIDIA A100-SXM4-80GB with 79.14GiB memory
[rank0]:[titan] 2025-05-28 21:11:34,788 - root - INFO - Model llama3 70B size: 33,851,922,432 total parameters
[rank0]:[titan] 2025-05-28 21:11:34,788 - root - INFO - No 'pipeline_parallel_split_points' provided. Here is the auto-generated split, which may be sub-optimal: ['layers.3', 'layers.7', 'layers.11', 'layers.15', 'layers.19', 'layers.23', 'layers.27', 'layers.31', 'layers.35', 'layers.39', 'layers.43', 'layers.47', 'layers.51', 'layers.55', 'layers.58'].
[rank0]:[titan] 2025-05-28 21:11:34,839 - root - INFO - PP rank 0 is building stage_idx 0 with start_layer None, stop_layer layers.3
[rank0]:[titan] 2025-05-28 21:11:34,839 - root - INFO - Applied full activation checkpointing to the model
[rank0]:[titan] 2025-05-28 21:11:34,839 - root - WARNING - Number of microbatches (2) is less than the total number of stages (16) which may result in a bubble in the pipeline.
[rank0]:[titan] 2025-05-28 21:11:34,839 - root - INFO - Using pipeline schedule GPipe with 2 microbatches and 16 stages.
[rank0]:[titan] 2025-05-28 21:11:37,625 - root - INFO - Peak FLOPS used for computing MFU: 3.120e+14
[rank0]:[titan] 2025-05-28 21:11:37,625 - root - INFO - CUDA memory usage for model: 8.97GiB(11.33%)
[rank0]:[titan] 2025-05-28 21:11:37,626 - root - INFO - Trainer is initialized with local batch size 2, global batch size 2, sequence length 4096, total steps 10 (warmup 2).
[rank0]:[titan] 2025-05-28 21:11:37,626 - root - INFO - Training starts at step 1.
[rank0]:[titan] 2025-05-28 21:13:37,816 - root - INFO - step:  1  loss: -1.0000  memory: 36.90GiB(46.63%)  tps: 4  tflops: 0.90  mfu: 0.29%
[rank0]:[titan] 2025-05-28 21:13:37,816 - root - INFO - Synchronizing and adjusting timeout for all ProcessGroups to 0:01:40
[rank0]:[titan] 2025-05-28 21:14:39,054 - root - INFO - step:  2  loss: -1.0000  memory: 42.78GiB(54.06%)  tps: 8  tflops: 1.81  mfu: 0.58%
[rank0]:[titan] 2025-05-28 21:15:40,293 - root - INFO - step:  3  loss: -1.0000  memory: 42.78GiB(54.06%)  tps: 8  tflops: 1.81  mfu: 0.58%
[rank0]:[titan] 2025-05-28 21:16:41,533 - root - INFO - step:  4  loss: -1.0000  memory: 42.78GiB(54.06%)  tps: 8  tflops: 1.81  mfu: 0.58%
[rank0]:[titan] 2025-05-28 21:17:42,765 - root - INFO - step:  5  loss: -1.0000  memory: 42.78GiB(54.06%)  tps: 8  tflops: 1.81  mfu: 0.58%
[rank0]:[titan] 2025-05-28 21:18:43,999 - root - INFO - step:  6  loss: -1.0000  memory: 42.78GiB(54.06%)  tps: 8  tflops: 1.81  mfu: 0.58%
[rank0]:[titan] 2025-05-28 21:19:45,234 - root - INFO - step:  7  loss: -1.0000  memory: 42.78GiB(54.06%)  tps: 8  tflops: 1.81  mfu: 0.58%
[rank0]:[titan] 2025-05-28 21:20:46,459 - root - INFO - step:  8  loss: -1.0000  memory: 42.78GiB(54.06%)  tps: 8  tflops: 1.81  mfu: 0.58%
[rank0]:[titan] 2025-05-28 21:21:47,679 - root - INFO - step:  9  loss: -1.0000  memory: 42.78GiB(54.06%)  tps: 8  tflops: 1.81  mfu: 0.58%
[rank0]:[titan] 2025-05-28 21:22:48,907 - root - INFO - step: 10  loss: -1.0000  memory: 42.78GiB(54.06%)  tps: 8  tflops: 1.81  mfu: 0.58%
[rank0]:[titan] 2025-05-28 21:22:48,922 - root - INFO - Sleeping 2 seconds for other ranks to complete
[rank0]:[titan] 2025-05-28 21:22:50,924 - root - INFO - Training completed
[rank0]:[titan] 2025-05-28 21:22:53,972 - root - INFO - Process group destroyed.

```

**Comment by lessw2020** [comment]
Hi @aahehehe I think there is a config issue. You said you are running two nodes for 16 gpus, but your .sh shown above only shows 8 gpus. 
~~~
+ NGPU=8
~~~
Can you double check this?

**Comment by lessw2020** [comment]
Hi @aahehehe  - I see - yes, that won't work b/c you are effectively creating a world size of 32 (2 nodes, 16 proc).

Can you clarify exactly what your setup is?  I mean I know you have 2 nodes of 8 gpus, but normally for multi-node we have slurm acting as the coordinator and that's how Titan then can 'see' all 16 gpus. 
Are you just running the standard bash run_training.sh?  That won't natively connect the two nodes...hence you will only see 8 gpus which seems to match with what is going on here.
Please let me know.  We can hook the two nodes together so torchrun can see thiings, but we have to add some additional instructions.

**Comment by aahehehe** [comment]
Hi @lessw2020 , thank you very much for your reply！yes，I run run_training.sh, and this is my setup. You mean I shouldn't use run_training.sh when I'm running with 2nodes of 8 gpus, and I noticed that there's a file called multinode_trainer.slurm, should I execute that?
```
set -ex

# use envs as local overrides for convenience
# e.g.
# LOG_RANK=0,1 NGPU=4 ./run_train.sh
NGPU=${NGPU:-"16"}
export LOG_RANK=${LOG_RANK:-0,15}
# CONFIG_FILE=${CONFIG_FILE:-"./torchtitan/models/llama3/train_configs/debug_model.toml"}
CONFIG_FILE=${CONFIG_FILE:-"./torchtitan/models/llama3/train_configs/llama3_70b.toml"}

overrides=""
if [ $# -ne 0 ]; then
    overrides="$*"
fi

TORCHFT_LIGHTHOUSE=${TORCHFT_LIGHTHOUSE:-"http://localhost:29510"}

PYTORCH_CUDA_ALLOC_CONF="expandable_segments:True" \
TORCHFT_LIGHTHOUSE=${TORCHFT_LIGHTHOUSE} \
torchrun --nnodes 2  --nproc_per_node=${NGPU} --node_rank 1 --master_addr 51.62.15.15 --master_port 29500  \
--local-ranks-filter ${LOG_RANK} --role rank --tee 3 \
-m torchtitan.train --job.config_file ${CONFIG_FILE} $overrides
```
